src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs

```diff
@@ -408,28 +410,35 @@ private void RemoveStalePools()
             if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
         }
 
+        private static string GetIdentityIfDefaultCredentialsUsed(bool defaultCredentialsUsed)
+        {
+            return defaultCredentialsUsed ? CurrentUserIdentityProvider.GetIdentity() : string.Empty;
+        }
+
         internal readonly struct HttpConnectionKey : IEquatable<HttpConnectionKey>
         {
             public readonly HttpConnectionKind Kind;
             public readonly string Host;
             public readonly int Port;
             public readonly string SslHostName;     // null if not SSL
             public readonly Uri ProxyUri;
+            public readonly string Identity;
 
-            public HttpConnectionKey(HttpConnectionKind kind, string host, int port, string sslHostName, Uri proxyUri)
+            public HttpConnectionKey(HttpConnectionKind kind, string host, int port, string sslHostName, Uri proxyUri, string identity)
             {
                 Kind = kind;
                 Host = host;
                 Port = port;
                 SslHostName = sslHostName;
                 ProxyUri = proxyUri;
+                Identity = identity;
             }
 
             // In the common case, SslHostName (when present) is equal to Host.  If so, don't include in hash.
             public override int GetHashCode() =>
                 (SslHostName == Host ?
-                    HashCode.Combine(Kind, Host, Port, ProxyUri) :
-                    HashCode.Combine(Kind, Host, Port, SslHostName, ProxyUri));
+                    HashCode.Combine(Kind, Host, Port, ProxyUri, Identity) :
+                    HashCode.Combine(Kind, Host, Port, SslHostName, ProxyUri, Identity));
```

Comment:
Did you measure if there's any impact to including Identity in the hash code?  From a correctness perspective, it's correct either way (Equals needs to include it, but GetHashCode needn't), and then it becomes a question of what's more expensive overall: factoring Identity into the hash code on every request, or potentially having some more collisions due to the same request being made with different identities.  I don't know which is better, but it's probably worth a few minutes experimenting.

Reply:
I see your point here, but it seems the actual performance will depend on different factors in prod environment, such as number of open connections on the same HttpClient, number of different identities being used for the same connection. Do you believe this micro-optimization can noticeably contribute to throughput considering that we are making network calls here?

Reply:
> Do you believe this micro-optimization can noticeably contribute to throughput considering that we are making network calls here?

For microservices, talking over loopback, etc... maybe, maybe not.  It's why I suggested measuring.

Reply:
@stephentoub  I measured performance of the following code for two GetHashCode versions. The first includes Identity in GetHashCode, the second doesn't.
```C#
for (var i = 0; i < 10000; i++)
{
  connections.TryAdd(keys[i % 10], "123");
}
```
Overall, including Identity increases performance ~1.63x, however this is an edge case where each operation causes a collision, in the real environment collision ratio will be less.

|    Method |     Mean |   Error |  StdDev |
|---------- |---------:|--------:|--------:|
| TryAddKey (Identity **hashed**) | 860.3 us | 4.19 us | 3.71 us |
| TryAddKey (Identity **not hashed**) | 1.400 ms | 0.0082 ms | 0.0073 ms |

Reply:
Thanks.  Can you share the benchmark?

Reply:
```C#
public class Program
    {
        static void Main(string[] args) => BenchmarkSwitcher.FromTypes(new[] { typeof(Test) }).Run(args);

        public class Test
        {
            private ConcurrentDictionary<HttpConnectionKey, string> connections = new ConcurrentDictionary<HttpConnectionKey, string>();
            private HttpConnectionKey[] keys = new HttpConnectionKey[10];

            [GlobalSetup]
            public void Setup()
            {
                for (var i = 0; i < keys.Length; i++)
                {
                    keys[i] = new HttpConnectionKey(HttpConnectionKind.Http, "testA", 1, "testB", null, $"DOMAIN/{i.ToString()}");
                }
            }

            [Benchmark]
            public void TryAddKey()
            {
                for (var i = 0; i < 10000; i++)
                {
                    connections.TryAdd(keys[i % 10], "123");
                }
            }
        }
    }

Reply:
> keys[i] = new HttpConnectionKey(HttpConnectionKind.Http, "testA", 1, "testB", null, $"DOMAIN/{i.ToString()}");

What are the results of your test if the identity is always the same rather than different for each addition?

Reply:
Currently, I have only 10 different identities and repeat TryAdd for 10000 times which causes 999 collisions per each identity. Do you want me to use the only identity with 9999 collisions?

Reply:
> Do you want me to use the only identity with 9999 collisions?

If you don't mind.  I'm just curious.  If the identity is the same, then everything will be a collision, at which point the extra hashing is pure overhead for zero gain.  I'd just like to understand what the numbers look like in that case.

Reply:
Actually, I just noticed you're doing lots of adds... that's not the interesting case.  Can you instead do lots of TryGetValues?

Reply:
Results for TryAdd with the single key is below. Will measure TryGet shortly.
```connections.TryAdd(keys[0], "123");```

|    Method |     Mean |   Error |  StdDev |
|---------- |---------:|--------:|--------:|
| TryAddKey (Identity hashed) | 868.9 us | 17.14 us | 22.88 us |
| TryAddKey (Identity not hashed) | 758.6 us | 2.19 us | 1.83 us |

Reply:
TryGet results for **10 keys**

|    Method |     Mean |     Error |    StdDev |
|---------- |---------:|----------:|----------:|
| TryGetKey (Identity **hashed**) | 579.7 us | 2.28 us | 2.13 us |
| TryGetKey (Identity **not hashed**) | 1.113 ms | 0.0075 ms | 0.0067 ms |

TryGet results for the **single key**

|    Method |     Mean |    Error |   StdDev |
|---------- |---------:|---------:|---------:|
| TryGetKey (Identity **hashed**) | 571.2 us | 3.64 us | 3.40 us |
| TryGetKey (Identity **not hashed**) | 486.5 us | 10.86 us | 10.16 us |

```C#
            [Benchmark]
            public void TryGetKey()
            {
                for (var i = 0; i < 10000; i++)
                {
                    connections.TryGetValue(keys[i % 10], out _);
                }
            }
```
```C#
            [Benchmark]
            public void TryGetKey()
            {
                for (var i = 0; i < 10000; i++)
                {
                    connections.TryGetValue(keys[0], out _);
                }
            }
```

Reply:
Thanks.  What does the setup look like for those tests?

Reply:
For the both of the tests Setup method is the same as below
```C#
            [GlobalSetup]
            public void Setup()
            {
                for (var i = 0; i < keys.Length; i++)
                {
                    keys[i] = new HttpConnectionKey(HttpConnectionKind.Http, "testA", 1, "testB", null, $"DOMAIN/{i.ToString()}");
                    connections.TryAdd(keys[i], "123");
                }
            }
```
The difference is only in the array size.

- 10 keys ```private HttpConnectionKey[] keys = new HttpConnectionKey[10];```
- 1 key ```private HttpConnectionKey[] keys = new HttpConnectionKey[1];```